enhancements/sig-policy/99-policy-placement-strategy/README.md

```diff
@@ -0,0 +1,269 @@
+    ought to be safe to remove it at this point if it's feasible and not overly complex to do so).
+
+- **Snippet (untested)**
+
+  ```golang
+  // common.go
+  type placementDecisionGetter struct {
+    c client.Client
+  }
+
+  func (pd placementDecisionGetter) List(selector labels.Selector, namespace string) ([]*clusterv1beta1.PlacementDecision, error) {
+    pdList := &clusterv1beta1.PlacementDecisionList{}
+    lopts := &client.ListOptions{
+      LabelSelector: selector,
+      Namespace:     namespace,
+    }
+
+    err := pd.c.List(context.TODO(), pdList, lopts)
+    pdPtrList := []*clusterv1beta1.PlacementDecision{}
+    for _, pd := range pdList.Items {
+      pdPtrList = append(pdPtrList, &pd)
+    }
+
+    return pdPtrList, err
+  }
+
+  func GetRolloutHandler(c client.Client, placement *clusterv1beta1.Placement) (*clusterv1alpha1.RolloutHandler, error) {
+    pdTracker := clusterv1beta1.NewPlacementDecisionClustersTracker(placement, placementDecisionGetter{c}, nil, nil)
+
+    return clusterv1alpha1.NewRolloutHandler(pdTracker)
+  }
+
+  var GetClusterRolloutStatus clusterv1alpha1.ClusterRolloutStatusFunc = func(clusterName string) clusterv1alpha1.ClusterRolloutStatus {
+    // (Phase 3 rollout logic goes here)
+
+    return clusterv1alpha1.ClusterRolloutStatus{
+      Status: "",             // initially determined by comparing the generation of the policy with the particular
+                              //   cluster's generation, resolving to either "ToApply" or "Succeed"
+      LastTransitionTime: "", // retrieved from the last transition time in the cluster's status
+    }
+  }
+  ```
+
+  ```golang
+  rolloutHandler := common.GetRolloutHandler(client, placement)
+  strategy, rolloutResult, err := rolloutHandler.GetRolloutCluster(clusterv1alpha1.All, common.GetClusterRolloutStatus)
+
+  // ... Use rolloutResults for propagation
+  ```
+
+#### Story 3 / Phase 3 (Stretch goal)
+
+As a system administrator, I want policies to be rolled out only when the policy on previous clusters show as
+"Compliant" and control their concurrency and priority.
+
+- **Summary**
+
+  - Only continue policy deployment once the previously deployed clusters show as "Compliant" after a configurable
+    timeout.
+
+- **Details**
+
+  - Update the `PlacementBinding` CRD to contain the `RolloutStrategy` struct. (See the
+    [`v1alpha1/RolloutStrategy`](https://github.com/open-cluster-management-io/api/blob/main/cluster/v1alpha1/types_rolloutstrategy.go))
+    Default to `All` if a strategy is not provided.
+  - Update the `governance-policy-propagator` controller to add a `open-cluster-management.io/root-policy-timestamp`
+    annotation to the replicated policy for each cluster indicating when the replicated policy was last applied.
+  - Store a `lastEvaluated` in the status of the replicated policies so that the `RolloutStatus` can be determined
```

Comment:
Could you please elaborate on this more?

I'm thinking about the case where the root policy is updated and the ConfigurationPolicy gets evaluated but the status doesn't change, how would this work? Also, if the root policy is updated but the generation of the ConfigurationPolicy doesn't change and its evaluation interval is like `2h` or `never` when compliant, this would not trigger an evaluation in time.

Other controllers also don't post when they last evaluated.

Reply:
We do provide the `LastEvaluated` in the ConfigurationPolicy, so I was thinking we can add logic to have the status-sync controller update the evaluation time when the `root-policy-timestamp` is greater than the last evaluated time, but I'd have to dig into when the `LastEvaluated` gets sent if it's not on every reconcile...
```golang
        // An ISO-8601 timestamp of the last time the policy was evaluated
        LastEvaluated string `json:"lastEvaluated,omitempty"`
```
https://github.com/open-cluster-management-io/config-policy-controller/blob/7e648171a8bd1ac051fe130f5539fbc36347907b/api/v1/configurationpolicy_types.go#L197 

Good point about other controllers, though. I'm not sure--I'll have to consider it more and look at our code. Maybe wipe the status on an update to let it repopulate or something?

Reply:
@dhaiducek after brainstorming with @gparvin since he has a good sense of the requirements from a core user, I'd like to propose the following:

* New policies are progressively propagated to clusters regardless of whether they are inform or enforce. The status of the root policy should indicate that cluster in later groups are pending rollout or some kind of other state.
* Existing policies that are updated should only have their `remediationAction: enforce` progressively applied. Take for instance a scenario where we have clusters A, B, and C and A is a canary cluster. An update to my policy that sets `remediationAction: enforce` would at first have `remedation: enforce` on cluster A and `remediationAction: inform` on clusters B and C. The root policy status needs to indicate that clusters B and C are pending enforcement. For example, their compliance state could be `NonCompliant - Pending Rollout`.
* PolicyAutomation (separate story) would have a new mode of `everyRollout` so can be triggered when a rollout fails as opposed to each cluster becoming noncompliant.


@dhaiducek and @JustinKuli , what do you think?

Reply:
I like the idea to switch to inform to prevent unintended updates. We could add the `RolloutStatus` to each cluster's status, but I'm not sure we should change the compliance state to NonCompliant during a rollout, especially if users have alerts set up or are active users of the dashboards--that could get unintentionally noisy? Maybe we just use the `Pending` state? But overall I think this is the right direction to having a controlled rollout.

Reply:
It's a fair point but at the same time, as a user, I would like to see the state of the policy on the cluster if there are inform policies applied even if they are pending rollout.

We could use the `Pending` state but include the noncompliant message so at least the information is there.

What do you think @dhaiducek and @gparvin?

Reply:
I prefer the compliance being NonCompliant when that reflects reality and "maybe" have a separate Remediation Pending status (new field).  Though there may be too many things we could do in this extra remediation status field:
- pending rollout
- remediation applied by policy controller
- automated remediation applied
- remediated by automation or user intervention
- no remediation needed

Reply:
@gparvin I'm unclear on when you say "when that reflects reality". Are you saying policies should continue to report based on the content it has at the time, or are you saying that if the policy is out-of-date, it should be considered NonCompliant? For the latter, I suppose I'm thinking about a case where it might have been a minor update to the policy where the ultimate compliance wouldn't have changed due to the update? But maybe there's a "Policy" best practice I'm missing there if that's the meaning.
And, if we set it to "NonCompliant" because a rollout is happening, how would setting it to "NonCompliant" affect Ansible automation? As @gparvin points out, we might need some way to distinguish "NonCompliant--this is out of date" vs "NonCompliant--this is broken"? In that case, should Ansible automation only act when the `RolloutStatus` is `Succeed`? (I'm crossing my fingers that gets updated to `Succeeded` 😆)
Interestingly, to the "too many things we could do" point, having an `enum` for status is apparently against the Kubernetes status conventions (though it seems `RolloutStatus` is already set as such), where "Status" should simply be "True", "False", or "Unknown": https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

Reply:
@dhaiducek, I think the point is that if we follow the idea that a progressive rollout to a policy update makes the enforce part progressive and nothing else, then the pending rollout clusters should still accurately report back their inform status.

Having an additional field in the root policy status, propagated policy status, or both indicating that a rollout is pending would be sufficient in my opinion.

For the PolicyAutomation front, I don't want us to get caught up on this since it's a very underutilized feature. For now, we could have an additional optional spec field on the PolicyAutomation such as `ignorePendingRollout`.

Reply:
Okay, that all sounds good to me, too. I've updated the enhancement with a point about setting the `remediationAction` to `inform`. I still need to check whether removing the status might work as a better alternative to the `lastEvaluated` status.